### ISSUE TITLE
No mechanism to communicate initialization status to init system

### ISSUE THREAD

**Issue by jcowgill** [gerbera/gerbera#152]
Gerbera has no mechanism to tell the init system if there was a failure during initialization or when it completes. Depending on the init system, this may allow services depending on gerbera to start before gerbera has finished initialization. On a related issue, init systems might report that gerbera starts successfully even though it didn't. Example with systemd:

```
# echo 'garbage' > /etc/gerbera/config.xml 
# systemctl start gerbera
# echo $?
0
# systemctl status gerbera
● gerbera.service - Gerbera Media Server
   Loaded: loaded (/lib/systemd/system/gerbera.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Sat 2017-09-16 19:37:03 BST; 2s ago
  Process: 30671 ExecStart=/usr/bin/gerbera -c /etc/gerbera/config.xml (code=exited, status=1/FAILURE)
 Main PID: 30671 (code=exited, status=1/FAILURE)

Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03    INFO: Gerbera UPnP Server version 1.0.0 - http://gerbera.io/
Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03    INFO: ===============================================================================
Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03    INFO: Gerbera is free software, covered by the GNU General Public License version 2
Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03    INFO: Copyright 2016-2017 Gerbera Contributors.
Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03    INFO: Gerbera is based on MediaTomb: Copyright 2005-2010 Gena Batsyan, Sergey Bostandzh
Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03    INFO: Loading configuration from: /etc/gerbera/config.xml
Sep 16 19:37:03 shenron gerbera[30671]: 2017-09-16 19:37:03   ERROR: Error parsing config file: /etc/gerbera/config.xml line 1:
Sep 16 19:37:03 shenron systemd[1]: gerbera.service: Main process exited, code=exited, status=1/FAILURE
Sep 16 19:37:03 shenron systemd[1]: gerbera.service: Unit entered failed state.
Sep 16 19:37:03 shenron systemd[1]: gerbera.service: Failed with result 'exit-code'.
```

Traditionally, this was implemented by forking after initialization as a signal to the init system that it was now started. This might work for checking configuration, but I think it will break libupnp's threading if done after upnp is initialized (among other problems). Probably the best way is to use `sd_notify`, but that will only work with systemd.

**Comment by v00d00** [comment]
I do want to add libsystemd support at some point.

Could you use start-stop-daemon's `--wait` to give gerbera a few seconds to init and die in the failure case?

**Comment by jcowgill** [comment]
Unfortunately Debian's start-stop-daemon doesn't have the `--wait` option. I think it's specific to OpenRC. For systemd, I think I can put an `ExecStartPost=/bin/sleep 1` in the service file. Waiting some arbitrary amount of time feels like a big hack though.

**Comment by v00d00** [comment]
The wait is absolutely a hack, but its the best we have for now 👎

**Comment by chiraag-nataraj** [comment]
@jcowgill Wait...the systemd example above clearly shows that systemd knows that gerbera failed. So if something depended on gerbera, it wouldn't be started. systemctl itself returns 0 because it found the unit file (aka, there was no error in _systemctl_). Using exit codes with systemd is probably a bad idea, and you should be using systemctl is-failed or systemctl is-active to figure out if a unit is active (interactively).

**Comment by jcowgill** [comment]
@chiraag-nataraj 
> Wait...the systemd example above clearly shows that systemd knows that gerbera failed. So if something depended on gerbera, it wouldn't be started.

This is incorrect. In the current situation (with `Type=simple`) systemd will not put gerbera into the "activating" state and instead will immediately make it "active" when you ask systemd to start it. It will immediately report that the service has started and proceed with starting any services which depend on it before it has a chance to fail. In my example it goes through this transition:
```
<request to start> => active => failed
```
Instead of this, we want systemd to wait until gerbera has finished starting up before marking the service active. This allows dependencies to work properly, and for systemctl to report any errors. The diagram should look like this:
```
<request to start> => activating => failed
```

> systemctl itself returns 0 because it found the unit file (aka, there was no error in systemctl) and that wouldn't change with Type=notify in the unit file (and using something like sd_notify).

systemctl does report errors (and returns status 1) if a service fails before becoming active. This will never happen with a `Type=simple` service though (for the reasons above) but should work with `Type=notify`.

> Using exit codes with systemd is probably a bad idea, and you should be using systemctl is-failed or systemctl is-active to figure out if a unit is active (interactively).

Firstly, my use of systemctl above is an example to illustrate the bug. Any other systemd API will produce similar issues.

Secondly, why is using the exit status of systemctl bad? It works fine for most other services and is documented to do so. This was the reason I noticed this bug - Debian packaging scripts use the output of systemctl to test if a service started successfully.

**Comment by chiraag-nataraj** [comment]
My bad. Thanks for the clarifications. What you're saying makes sense :)
